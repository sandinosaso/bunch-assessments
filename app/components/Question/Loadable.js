/**
 *
 * Asynchronously loads the component for Question
 *
 */

import Loadable from 'react-loadable';

export default Loadable({
  loader: () => import('./index'),
  loading: () => null,
});
