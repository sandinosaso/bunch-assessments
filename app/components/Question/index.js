/**
*
* Question
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Field } from 'redux-form/immutable';
import { QuestionPropType } from '../../lib/PropTypesValues';
import CustomRadioInput from '../CustomRadioInput';

const Input = styled.input`
  padding: 0.5em;
  margin: 0.5em;
  color: palevioletred;
  background: papayawhip;
  border: none;
  border-radius: 3px;
`;

const requiredValidation = (value) => value ? undefined : 'Question is required. Please choose one option';

const renderField = ({ input, label, type, meta: { touched, error } }) =>
(<div>
  <label htmlFor>
    {label}
  </label>
  <div>
    <Input {...input} type={type} placeholder={label} />
    {touched &&
      error &&
      <span>
        {error}
      </span>}
  </div>
</div>);

const Question = (props) => {
  const { data: question, style } = props;
  if (question.type === 'radio') {
    return (
      <Field
        name={question.name}
        component={CustomRadioInput}
        heading={question.label}
        options={question.options}
        validate={[
          requiredValidation,
        ]}
        style={style}
      />
    );
  }

  return (<Field
    name={question.name}
    type={question.type}
    component={renderField}
    label={question.label}
  />);
};

renderField.propTypes = {
  input: PropTypes.object,
  label: PropTypes.string,
  type: PropTypes.string,
  meta: PropTypes.object,
};

Question.propTypes = {
  data: QuestionPropType,
  style: PropTypes.object,
};

export default Question;
