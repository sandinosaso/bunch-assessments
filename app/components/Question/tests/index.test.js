import React from 'react';
import { shallow } from 'enzyme';

import Question from '../index';

describe('<Question />', () => {
  it('Renders Question content when passing all props', () => {
    const data = {
      name: 'Question Name',
      label: 'Question Label',
      input: {
      },
      type: 'radio',
      options: [{ label: 'Option 1', value: 1 }, { label: 'Option 2', value: 2 }],
    };
    const wrapper = shallow(
      <Question
        data={data}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
