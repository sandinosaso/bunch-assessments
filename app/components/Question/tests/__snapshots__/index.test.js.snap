// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Question /> Renders Question content when passing all props 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Field
    component={[Function]}
    heading="Question Label"
    name="Question Name"
    options={
        Array [
            Object {
              "label": "Option 1",
              "value": 1,
            },
            Object {
              "label": "Option 2",
              "value": 2,
            },
          ]
    }
    style={undefined}
    validate={
        Array [
            [Function],
          ]
    }
/>,
  "nodes": Array [
    <Field
      component={[Function]}
      heading="Question Label"
      name="Question Name"
      options={
            Array [
                  Object {
                    "label": "Option 1",
                    "value": 1,
                  },
                  Object {
                    "label": "Option 2",
                    "value": 2,
                  },
                ]
      }
      style={undefined}
      validate={
            Array [
                  [Function],
                ]
      }
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Question
        data={
                Object {
                        "input": Object {},
                        "label": "Question Label",
                        "name": "Question Name",
                        "options": Array [
                          Object {
                            "label": "Option 1",
                            "value": 1,
                          },
                          Object {
                            "label": "Option 2",
                            "value": 2,
                          },
                        ],
                        "type": "radio",
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "data": Object {
            "input": Object {},
            "label": "Question Label",
            "name": "Question Name",
            "options": Array [
              Object {
                "label": "Option 1",
                "value": 1,
              },
              Object {
                "label": "Option 2",
                "value": 2,
              },
            ],
            "type": "radio",
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Field
          component={[Function]}
          heading="Question Label"
          name="Question Name"
          options={
                    Array [
                              Object {
                                "label": "Option 1",
                                "value": 1,
                              },
                              Object {
                                "label": "Option 2",
                                "value": 2,
                              },
                            ]
          }
          style={undefined}
          validate={
                    Array [
                              [Function],
                            ]
          }
/>,
        "_debugID": 2,
        "_renderedOutput": <Field
          component={[Function]}
          heading="Question Label"
          name="Question Name"
          options={
                    Array [
                              Object {
                                "label": "Option 1",
                                "value": 1,
                              },
                              Object {
                                "label": "Option 2",
                                "value": 2,
                              },
                            ]
          }
          style={undefined}
          validate={
                    Array [
                              [Function],
                            ]
          }
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Question
    data={
        Object {
            "input": Object {},
            "label": "Question Label",
            "name": "Question Name",
            "options": Array [
              Object {
                "label": "Option 1",
                "value": 1,
              },
              Object {
                "label": "Option 2",
                "value": 2,
              },
            ],
            "type": "radio",
          }
    }
/>,
}
`;
